mono/mini/aot-runtime.c

```diff
@@ -2537,6 +2540,8 @@ load_aot_module (MonoAssemblyLoadContext *alc, MonoAssembly *assembly, gpointer
 	} else {
 		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_AOT, "AOT: image '%s' found.", found_aot_name);
 	}
+
+	load_container_amodule (alc);
```

Comment:
This seems like a very indirect way to fix the problem.

Also - if you do the load here, can you delete it from `mono_aot_get_method`

More generally, I don't understand what's going on with `container_assm_name`:
1. So the first AOT image we see that's tagged by `MONO_AOT_FILE_FLAG_EAGER_LOAD` is the container assembly? what happens if we see a second one?
2. If it has to be loaded early before other AOT images, why do we load it indirectly in `load_aot_module` every time?  If we know we have to load it early, just load it during startup.
3. If it's an AOT image full of dedup code, where does its .dll or .exe come from? or are we dumping everything into the AOT image of the main assembly?  If so, why all this stuff with a `container_assm_name`?
4. There's something clever going on with lock-free manipulation of `container_assm_name` to pass ownership of it from the code that assigns it to `load_container_amdoule` - but there are no atomic operations and no locking - is it protected by `mono_aot_lock` - I don't think so - the old caller in `mono_aot_get_method` doesn't look like it takes that lock.

Reply:
Capturing some comments from offline:

The container assembly is a dummy assembly for AOT deduplication.  for wasm it's `aot-instances.dll` - the assembly is empty, but it's AOT image is full of dedup'd instances.  (Architecturally we need an assembly to hang an image from. we can't just load a bare AOT image)

Embedders call `mono_aot_register_module` with a `MONO_AOT_FILE_FLAG_EAGER_LOAD` flag to register the dedup container.  There is at most one.

What we really want is just to load it after corlib is loaded but before any managed loader hooks are set up.  Alternately, we want to load the dedup container without firing managed loader hooks.   (Because that ends up being a codepath that's not evident to the AOT compiler, so we end up trying to JIT code while in FullAOT mode).

---

So we're going to make a loader flag for `load_container_amodule` that says "don't run managed hooks when loading this assembly" (rationale: the aot instances assembly is an implementation detail, and is empty in any case)